src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexCharClass.cs

```diff
@@ -1279,54 +1279,96 @@ private void Canonicalize()
+                {
+                    rangelist.Sort((x, y) => x.First.CompareTo(y.First));
 
-                bool done = false;
-                int i, j;
+                    bool done = false;
+                    int j = 0;
 
-                for (i = 1, j = 0; ; i++)
-                {
-                    char last;
-                    for (last = rangelist[j].Last; ; i++)
+                    for (int i = 1; ; i++)
                     {
-                        if (i == rangelist.Count || last == LastChar)
+                        char last;
+                        for (last = rangelist[j].Last; ; i++)
                         {
-                            done = true;
-                            break;
+                            if (i == rangelist.Count || last == LastChar)
+                            {
+                                done = true;
+                                break;
+                            }
+
+                            SingleRange currentRange;
+                            if ((currentRange = rangelist[i]).First > last + 1)
+                            {
+                                break;
+                            }
+
+                            if (last < currentRange.Last)
+                            {
+                                last = currentRange.Last;
+                            }
                         }
 
-                        SingleRange currentRange;
-                        if ((currentRange = rangelist[i]).First > last + 1)
+                        rangelist[j] = new SingleRange(rangelist[j].First, last);
+
+                        j++;
+
+                        if (done)
                         {
                             break;
                         }
 
-                        if (last < currentRange.Last)
+                        if (j < i)
                         {
-                            last = currentRange.Last;
+                            rangelist[j] = rangelist[i];
                         }
                     }
 
-                    rangelist[j] = new SingleRange(rangelist[j].First, last);
-
-                    j++;
+                    rangelist.RemoveRange(j, rangelist.Count - j);
+                }
 
-                    if (done)
+                // If the class now represents a single negated character, but does so by including every
+                // other character, invert it to produce a normalized form recognized by IsSingletonInverse.
+                if (!_negate && _subtractor is null && (_categories is null || _categories.Length == 0))
+                {
+                    if (rangelist.Count == 2)
                     {
-                        break;
+                        // There are two ranges in the list.  See if there's one missing element between them.
+                        if (rangelist[0].First == 0 &&
+                            rangelist[0].Last == (char)(rangelist[1].First - 2) &&
```

Comment:
Is it possible for `rangelist[1].First - 2` to "wrap around" or "underflow" at all here (for example if it had the value `\u0001`? And if it does - would that be a problem?

Reply:
It shouldn't be possible, as the only way that could happen is if the two ranges were contiguous, but if they were contiguous, they should have been merged as part of the normalization earlier.  Even if that didn't happen, however, we'd still be fine: this just wouldn't match and we'd move on with our lives :)  In short, we're looking for a very specific pattern here, and any deviation from that and we just won't make any changes.

Reply:
Ok - I didn't know if we got into this case with a pattern like:

```
@"[\u0000-\uFFFE\u0001-\uFFFF]"
```

Where `rangelist[1].First - 2` would be `\u0001 - 2` which is `\uFFFE`.

Reply:
I've added that as a test.  Thanks.